src/common/numbers/integer.cpp

```diff
@@ -0,0 +1,424 @@
+    }
+
+    for (auto it = stepSizes.crbegin(); it != stepSizes.crend(); ++it) {
+      auto [i, stepSize] = *it;
+      Integer step{stepSize};
+      while (step <= remainder) {
+        remainder -= step;
+        result = result + i;
+        ensure(remainder.positive()); // We don't support fractions, to we should always stop exactly at 0
+      }
+    }
+  } else {
+    auto step = o.absolute();
+
+    while (!remainder.slices.empty()) {
+      remainder -= step;
+      ++result;
+      ensure(remainder.positive()); // We don't support fractions, to we should always stop exactly at 0
+    }
+  }
+
+  result._positive = this->positive() == o.positive() || result.slices.empty();
+  return result;
+}
+
+Integer Integer::operator%(const Integer &o) const {
+  ensure(o != 0); // division by zero is undefined
+
+  if (slices.empty()) { // Zero divided by anything is always zero
+    return *this;
+  }
+
+  Integer remainder{this->slices, true};
+
+  if (o.slices.size() == 1) {
+    // Optimizing this common scenario
+    auto base = o.slices[0];
+    std::vector<intmax_t> stepSizes;
+    stepSizes.push_back(base);
+
+    constexpr auto max = std::numeric_limits<intmax_t>::max() / 2;
+    for (intmax_t i = 2; i <= max && (*this > stepSizes.back()); i *= 2) {
+      stepSizes.push_back(i * base);
+    }
+
+    for (auto it = stepSizes.crbegin(); it != stepSizes.crend(); ++it) {
+      auto stepSize = *it;
+      Integer step{stepSize};
+      while (step <= remainder) {
+        remainder -= stepSize;
+      }
+    }
+  } else {
+    auto step = o.absolute();
+    while (step <= remainder) {
+      remainder -= step;
+    }
+  }
+
+  remainder._positive = this->positive() == o.positive() || remainder.slices.empty();
+  return remainder;
+}
+
+Integer Integer::operator+(intmax_t value) const {
+  if (value == 0) return *this;
+  if (slices.empty()) return Integer{value};
+
+  auto sameSign = (value >= 0 && this->positive()) || (value < 0 && !this->positive());
+  auto absValue = std::abs(value);
+  if (sameSign && absValue + slices[0] < 1000000000) {
+    Integer result{*this};
+    result.slices[0] += absValue;
+    return result;
+  }
+
+  return *this + Integer{value};
+}
+
+Integer Integer::operator*(intmax_t value) const {
+  switch (value) {
+  case 0:
+    return Integer{0};
+  case 1:
+    return *this;
+  case -1:
+    return Integer{slices, !_positive};
+  default:
+    return *this * Integer(value);
+  }
+}
+
+Integer Integer::power(const Integer &exponent) const {
+  ensure(*this != 0 || exponent != 0); // zero ^ zero is undefined
+  ensure(exponent.positive());         // Haven't implemented this yet
+
+  Integer result{1};
+
+  for (Integer i{0}; i < exponent; ++i) {
+    result *= *this;
+  }
+
+  return result;
+}
+
+bool Integer::operator<(const Integer &o) const {
+  if (this->positive() != o.positive()) {
+    return this->positive() < o.positive();
+  }
+
+  auto bitComparison = compareSlices(this->slices, o.slices);
```

Comment:
Should be called `sliceComparison`
```suggestion
  auto sliceComparison = compareSlices(this->slices, o.slices);
```